src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/SpeakerWorkerForDumpsService.java

```diff
@@ -0,0 +1,146 @@
+     */
+    private final Map<String, List<ChunkedSpeakerDataResponse>> messagesChains;
+
+    public SpeakerWorkerForDumpsService(@NonNull SpeakerCommandForDumpsCarrier carrier,
+                                        int chunkedMessagesExpirationMinutes) {
+        this.carrier = carrier;
+        messagesChains = new PassiveExpiringMap<>(chunkedMessagesExpirationMinutes, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * Sends command to speaker.
+     *
+     * @param key     unique operation's key.
+     * @param request command to be executed.
+     */
+    public void sendCommand(@NonNull String key, @NonNull CommandMessage request) throws PipelineException {
+        log.debug("Got a request from hub bolt {}", request);
+        keyToRequest.put(key, request);
+        carrier.sendCommand(key, request);
+    }
+
+    /**
+     * Processes received response and forwards it to the hub component.
+     *
+     * @param key      operation's key.
+     * @param response response payload.
+     */
+    public void handleResponse(@NonNull String key, @NonNull SpeakerDataResponse response)
+            throws PipelineException {
+        log.debug("Got a response from speaker {}", response);
+        CommandMessage pendingRequest = keyToRequest.remove(key);
+        if (pendingRequest != null) {
+            carrier.sendResponse(key, response.getData());
+        }
+    }
+
+    /**
+     * Processes received chunked responses, combines them and forwards it to the hub component.
+     */
+    public void handleChunkedResponse(String key, ChunkedSpeakerDataResponse response) throws PipelineException {
+        log.debug("Got chunked response from speaker {}", response);
+        chunkedMessageIdsPerRequest.computeIfAbsent(key, mappingFunction -> new HashSet<>());
+        Set<String> associatedMessages = chunkedMessageIdsPerRequest.get(key);
+        if (!associatedMessages.add(response.getMessageId())) {
+            log.debug("Skipping chunked message, it is already received: {}", response);
+            return;
+        }
+
+        messagesChains.computeIfAbsent(key, mappingFunction -> new ArrayList<>());
+        List<ChunkedSpeakerDataResponse> chain = messagesChains.get(key);
+        if (response.getTotalMessages() != 0) {
+            chain.add(response);
+        }
+
+        if (chain.size() == response.getTotalMessages()) {
+            completeChunkedResponse(key, chain);
+        }
+    }
+
+    private void completeChunkedResponse(String key, List<ChunkedSpeakerDataResponse> messages)
+            throws PipelineException {
+
+        CommandMessage pending = keyToRequest.remove(key);
+        messagesChains.remove(key);
+        chunkedMessageIdsPerRequest.remove(key);
+
+        if (pending != null && pending.getData() != null) {
+            if (pending.getData() instanceof DumpRulesForFlowHsRequest) {
+                List<FlowDumpResponse> responses = messages.stream()
```

Comment:
since code is very simular for each type, perhaps the common method can be used with lambda as an argument

Reply:
well noted,
Fixed